### ISSUE TITLE
Parallel.ForEach on Mac seems to have no limit

### ISSUE THREAD

**Issue by chadwackerman** [dotnet/coreclr#5232]
I have a simple app that spins up command line executables via Microsoft.DotNet.Cli.Utils Command.Create()/Execute().

On Windows it seems to do the right thing. On an 8 core machine it creates 8-14 processes.

On a dual-core Mac it created over 90 processes before the machine started to tip over. Most showed no scheduling at all (not even 0%) in Activity Monitor.

**Comment by stephentoub** [comment]
If you don't specify a MaxDegreeOfParallelism via a ParallelOptions passed to Parallel.ForEach, there is no maximum.  ForEach will essentially rely on the thread pool to manage the number of threads involved, by creating a task which, when executed, will create another task, and so on.  A task will run one or more iterations, and then if there's still work to do, will queue a replica of itself and go away.  In this fashion, Parallel.ForEach lets the thread pool determine how many tasks should be run concurrently.  However, the thread pool also has a heuristic that'll inject more threads if all workers appear to be blocked, so if the body of your loop is just blocking waiting for some other process, it's likely that the thread pool will continue to inject threads at a rate of about a thread a second.

If there is a difference between Windows and OSX in your app, I doubt it has to do with Parallel.ForEach itself; rather, I expect the actual work being performed ends up being different in some way, such that the work is completing more quickly on Windows, such that the pool isn't seeing things as blocked as much and isn't introducing as many threads.

**Comment by chadwackerman** [comment]
The Mac is indeed a slower machine but it's pretty basic file processing on SSD. It's the heuristic that I'm questioning. The Mac spins up files sensibly until the last one gets no CPU. Then it creates a storm of starved processes, much like you predict -- but not what I'd expect.

Some of these processes are CPU bound, others are network bound, a few are potentially disk bound. On Windows I just let 'er rip. On Mac the machine grinds to an unusable halt and takes a significant amount of time to recover. Is this really expected behavior?

I'm familiar with the MaxDegreeOfParallelism option however I'm not sure there's an optimal number for my workload. The dotnet version is faster than Python (implying a dynamic limit smarter than Pool() has a benefit) however the Mac behavior seems to have sharper edges than I'd expect.

**Comment by stephentoub** [comment]
> Is this really expected behavior?

What is the body of the loop doing?  Just launching a process and then blocking waiting for that process to complete?  If so, Parallel.ForEach has no knowledge of the other processes: all the thread pool sees is that it's blocked waiting for something, and even if it was able to monitor CPU usage by its own process, network usage by its own process, etc., that wouldn't matter, as it's blocking waiting for something external to the process.  From the thread pool's perspective, it can't tell the difference between those threads being blocked on something that will complete soon and those threads being deadlocked, in which case it wants to introduce more threads in order to unblock them.

Without knowing more about the workload, I can't really speak to why you're seeing a difference between Windows and OSX here; I expect it has something to do with how processes are handled on those machines.  But, regardless, the heuristics employed by Parallel.ForEach in this case are unlikely to be what you want... I'm actually a bit surprised you're getting decent behavior on Windows.  You'd be better off picking some upper bound you're comfortable with and setting that as the MaxDegreeOfParallelism.

**Comment by chadwackerman** [comment]
Thank you for the detailed response. I researched the history of this going back to .NET 2.0 SP1 when the default changed from 25 threads per CPU to 250 per a Joe Duffy blog post. I've also seen quotes of 100/core, 20/core (in source code) and an assortment of other numbers where people are confused with the ASP.NET pool.

I also poked around in Mono to see what they did -- the default was 1 per CPU at one point!

Windows does a particularly amazing job with threads and my hunch is that you may need to revisit this at some point on other platforms. By way of example the default file handles on Mac is 256 per process which is basically a DOS-era number.

Thanks again for your time and info. Feel free to close this.

**Comment by stephentoub** [comment]
Thanks, @chadwackerman.
